File: android/SceneWorld001/app/src/main/java/z3d/display/particle/Display3DParticle.java
package z3d.display.particle;

import android.opengl.GLES20;
import android.util.Log;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.FloatBuffer;
import java.util.List;

import z3d.base.Object3D;
import z3d.base.Scene_data;
import z3d.core.Context3D;
import z3d.display.Display3D;
import z3d.display.particle.ctrl.TimeLine;
import z3d.material.DynamicConstItem;
import z3d.material.DynamicTexItem;
import z3d.material.TexItem;
import z3d.program.Shader3D;
import z3d.vo.Float32Array;
import z3d.vo.Matrix3D;
import z3d.vo.Vector3D;


public class Display3DParticle extends Display3D {

    public boolean visible;
    public TimeLine timeline;
    public Shader3D shader3D;
    public Matrix3D modeMatrix;
    protected float _time;
    private float _beginTime;
    public ParticleData data;
    public Matrix3D bindMatrix;
    public Vector3D bindVecter3d;
    public Vector3D bindScale;
    public Matrix3D invertBindMatrix;
    public Matrix3D groupMatrix;
    protected Matrix3D  _rotationMatrix;
    public Matrix3D  modelMatrix;
    public boolean isInGroup ;
    public Vector3D groupPos;
    public Vector3D groupScale;
    public Vector3D groupRotation;

    private String  TAG="Display3DParticle";
    public Display3DParticle()
    {
        super(null);

        this._time=0;
    }
    public void setBind(Vector3D $pos,Matrix3D  $rotation,Vector3D  $scale,Matrix3D  $invertRotation, Matrix3D $groupMatrix)
    {
        this.bindVecter3d = $pos;
        this.bindMatrix = $rotation;
        this.bindScale = $scale;
        this.invertBindMatrix = $invertRotation;
        this.groupMatrix = $groupMatrix;
    }
    public void setTimeLine(TimeLine $tl) {
        this.timeline = $tl;
        this._beginTime = $tl.beginTime;
    }
    public void onCreated()
    {

    }
    public void updateTime(float t)
    {


        this._time = t - this._beginTime;
        this._time += this.data._delayedTime; //加上延时
        this.timeline.updateTime(t);

        this.visible = this.timeline.visible;


        this.posMatrix3d.identity();
        this.posMatrix3d.prependScale(this.scaleX * 0.1f * this.bindScale.x * this.data.overAllScale,
                this.scaleY * 0.1f * this.bindScale.y * this.data.overAllScale,
                this.scaleZ * 0.1f * this.bindScale.z * this.data.overAllScale);

        this.timeline.updateMatrix(this.posMatrix3d, this);

    }
    public void update()
    {

        if (this.visible&& this.data!=null&&this.scene3d!=null){
            if( this.data.materialParam!=null&&this.shader3D!=null){
                Context3D ctx=this.scene3d.context3D;
                ctx.setProgame(this.shader3D.program);
                ctx.setBlendParticleFactors(this.data._alphaMode);
                ctx.cullFaceBack(this.data.materialParam.material.backCull);
                this.updateMatrix();
                this.setMaterialVc();
                this.setMaterialTexture();
                this.setVc();
                this.setVa();
                this.resetVa();

            }
        }
    }
    public void  setViewCamModeMatr3d()
    {

    }
    public void  setMaterialTexture()
    {
        Context3D ctx=this.scene3d.context3D;
        List<TexItem> texVec  = this.data.materialParam.material.texList;
        for (int i   = 0; i < texVec.size(); i++) {
            TexItem texItem=texVec.get(i);
            if (texItem.isDynamic) {
                continue;
            }
            ctx.setRenderTexture(this.shader3D,texItem.name,texItem.textureRes.textTureInt,texItem.get_id());
        }
        List<DynamicTexItem>  texDynamicVec  = this.data.materialParam.dynamicTexList;
        for (int i   = 0; i < texDynamicVec.size(); i++) {
            DynamicTexItem dynamicTexItem=  texDynamicVec.get(i);
          TexItem texItem=dynamicTexItem.target;
          if(dynamicTexItem.hasTextureRes()){
              ctx.setRenderTexture(this.shader3D,texItem.name,dynamicTexItem.getTexture(),texItem.get_id());
          }


        }
    }
    public void  reset()
    {

    }
    public void  setVa()
    {

    }
    public void  setVc()
    {

    }
    public void  resetVa()
    {

    }
    public void  setMaterialVc()
    {

        if (this.data.materialParam==null) {
            return;
        }
        List<DynamicConstItem> dynamicConstList= this.data.materialParam.dynamicConstList;

        float t= Math.floorMod ((long) this._time , (long) (Scene_data.frameTime * this.data._life));
        for (int i = 0; i < dynamicConstList.size(); i++) {
        dynamicConstList.get(i).update(t);
        }
        if(this.data.materialParam.material.fcNum <= 0){
            return;
        }
        t = t * this.data.materialParam.material.timeSpeed;
        this.data.materialParam.material.update(t);
        Float32Array fcData= this.data.materialParam.material.fcData;
        Context3D ctx=this.scene3d.context3D;
        int fcNum=this.data.materialParam.material.fcNum;
        fcData.put(0,0.0f);
        fcData.put(1,1.0f);
        fcData.put(2,0.0f);
        fcData.put(3,0.2f);
        ctx.setVc4fv(this.shader3D,"fc",fcNum, fcData.verBuff);


    }


}

File: metalios/iosgl/iosgl/bt/display/particle/Display3DParticle.m
//
//  Display3DParticle.m
//  iosgl
//
//  Created by zhao on 23/2/2020.
//  Copyright © 2020 zhao. All rights reserved.
//

#import "Display3DParticle.h"
#import "ParticleData.h"
#import "DynamicConstItem.h"
#import "DynamicTexItem.h"
#import "Context3D.h"
#import "Shader3D.h"
#import "Scene3D.h"
#import "Scene_data.h"
#import "Vector3D.h"
#import "Matrix3D.h"
#import "ParticleMetalType.h"

@interface Display3DParticle()
@property (nonatomic, strong)  TimeLine*  timeline;
@end

@implementation Display3DParticle

- (instancetype)init:value
{
    self = [super init:value];
    if (self) {
        self._time=0;
        self.visible=YES;
    }
    return self;
}
-(void)onCreated;
{
}
 

-(void)setBind:(Vector3D*)pos rotation:(Matrix3D*)rotation scale:(Vector3D*)scale invertRotation:(Matrix3D*)invertRotation groupMatrix:(Matrix3D*)groupMatrix;
{
    self.bindVecter3d = pos;
    self.bindMatrix = rotation;
    self.bindScale = scale;
    self.invertBindMatrix = invertRotation;
    self.groupMatrix = groupMatrix;
}


-(void)updateTime:(float)t;
{
    Display3DParticle* this=self;
    this._time = t - this.beginTime;
    [this.timeline updateTime:t];
    this.visible = this.timeline.visible;
    [this.posMatrix3d identity];
    [this.posMatrix3d prependScale:this.scaleX*0.1* this.bindScale.x y:this.scaleY*0.1* this.bindScale.y z:_scaleZ*0.1* this.bindScale.z];
    [this.timeline updateMatrix:self.posMatrix3d particle:this];
    
  //  NSLog(@"%f  %f   %f",self.bindVecter3d.x,self.bindVecter3d.y,self.bindVecter3d.z);
    
    
}
-(void)updateMatrix;
{
    if (!self.bindMatrix){
        return;
    }
    [self.modeMatrix identity];
    [self.modeMatrix append:self.posMatrix3d];
    [self.modeMatrix append:self.bindMatrix];
    
 
    
    [self.modeMatrix appendTranslation:self.bindVecter3d.x y:self.bindVecter3d.y z:self.bindVecter3d.z];
}

-(void)update;
{
    if(self.visible ){
        if ( self.data.materialParam){
          
            self.shader3d=self.data.materialParam.shader;
            
            [self.shader3d mtlSetProgramShader];
          
            [self updateMatrix];
            [self setMaterialVc];
            [self setMaterialTexture];
            [self setVc];
            [self setVa];
            [self resetVa];
        }
        
    }
}
/*
 设置基础透视，镜头，模型矩阵
 */
-(void)setViewCamModeMatr3d;
{
 
  
    Camera3D* cam3D=self.scene3D.camera3D;
    
//    [ctx setVcMatrix4fv:self.shader3d name:"viewMatrix" data:cam3D.viewMatrix.m];
//    [ctx setVcMatrix4fv:self.shader3d name:"camMatrix" data:cam3D.camMatrix3D.m];
//    [ctx setVcMatrix4fv:self.shader3d name:"modeMatrix" data:self.modeMatrix.m];

    /*
    id<MTLRenderCommandEncoder> renderEncoder=self.scene3D.context3D.renderEncoder;
    
    ParticleMetalMatrixData matrixList = {[cam3D.viewMatrix getMatrixFloat4x4], [cam3D.camMatrix3D getMatrixFloat4x4], [self.modeMatrix getMatrixFloat4x4]};
  
   [renderEncoder setVertexBytes:&matrixList
                          length:sizeof(matrixList)
                         atIndex:2];
    */
    

}

-(void)setMaterialTexture;
{
    Context3D *ctx=self.scene3D.context3D;
    NSArray<TexItem*>* texVec  = self.data.materialParam.material.texList;
    for (int i   = 0; i < texVec.count; i++) {
        if (texVec[i].isDynamic) {
            continue;
        }
        [ctx setRenderTexture:self.data.materialParam.shader name:texVec[i].name texture:  texVec[i].textureRes.textTureLuint level:0];
    }
    NSArray<DynamicTexItem*>* texDynamicVec  =( NSArray<DynamicTexItem*>*) self.data.materialParam.dynamicTexList;
    for (int i   = 0; i < texDynamicVec.count; i++) {
        TexItem* texItem=texDynamicVec[i].target;
//        [ctx setRenderTexture:self.data.materialParam.shader name:texDynamicVec[i].target.name  texture:texDynamicVec[i].texture level:texItem.id];
         
        id<MTLRenderCommandEncoder> renderEncoder=self.scene3D.context3D.renderEncoder;
        
      
        [renderEncoder setFragmentTexture:texDynamicVec[i].mtltexture
                                  atIndex:texItem.id];
    }
    
}
-(void)reset;
{
    [self.timeline reset];
    [self updateTime:0];
}
 
-(void)inverBind;
{
    if(self.invertBindMatrix){
        
    }
    
}
-(void)setVc;
{
}
-(TimeLine*)timeLine;
{
    return  _timeline;
}
-(void)setTimeLine:(TimeLine*)value;
{
    _timeline=value;
 
    self.beginTime=_timeline.beginTime;
  
}
-(void)setMaterialVc;
{
    Display3DParticle* this=self;
    if (!this.data.materialParam) {
        return;
    }
    NSMutableArray<DynamicConstItem*>* dynamicConstList= this.data.materialParam.dynamicConstList;
     float t= fmod (self._time , [Scene_data default].frameTime * this.data._life);
    for (int i = 0; i < dynamicConstList.count; i++) {
        [dynamicConstList[i] update:t];
    }
    if(this.data.materialParam.material.fcNum <= 0){
        return;
    }
    t = t * this.data.materialParam.material.timeSpeed;
    [this.data.materialParam.material update:t];
//    Context3D *ctx=self.scene3D.context3D;
    NSMutableArray<NSNumber*>*   fcData= this.data.materialParam.material.fcData;
    GLfloat fcDataGlArr[fcData.count];
    for (int i=0; i<fcData.count; i++) {
        fcDataGlArr[i]=fcData[i].floatValue;
    }
//    [ctx setVc4fv:this.data.materialParam.shader name:"fc" data:fcDataGlArr len:this.data.materialParam.material.fcNum];
    
    
//    GLfloat attrArr[4];
//    attrArr[0]=1;
//    attrArr[1]=1;
//    attrArr[2]=0;
//    attrArr[3]=1;
    if(fcData.count!=4){
        NSLog(@"出错setMaterialVc%lu",fcData.count);
    }
    
 
    
    
    
    id<MTLRenderCommandEncoder> renderEncoder=self.scene3D.context3D.renderEncoder;
    [renderEncoder setFragmentBytes:fcDataGlArr length:sizeof(fcDataGlArr) atIndex:[self getFcDataIdx]];
}
-(NSUInteger)getFcDataIdx;
{
    return 1; //FC数据的位置，不同特效不一样
}

-(void)setVa;
{
}
-(void)resetVa;
{
}


@end
